pkg/cloudprovider/aws/fleet/packing/packing.go

```diff
@@ -38,19 +42,141 @@ type Packings struct {
 	InstanceTypeOptions []string
 }
 
-func NewPacker(ec2 ec2iface.EC2API) *PodPacker {
-	return &PodPacker{ec2: ec2}
+// NewPacker returns a Packer implementation
+func NewPacker(ec2 ec2iface.EC2API) Packer {
+	return &podPacker{ec2: ec2}
 }
 
 // Pack returns the packings for the provided pods. Computes a set of viable
 // instance types for each packing of pods. Instance variety enables EC2 to make
-// better cost and availability decisions.
-func (p *PodPacker) Pack(ctx context.Context, pods []*v1.Pod) ([]*Packings, error) {
-	zap.S().Debugf("Successfully packed %d pod(s) onto %d node(s)", len(pods), 1)
-	return []*Packings{
-		{
-			InstanceTypeOptions: []string{"m5.large"}, // TODO, prioritize possible instance types
-			Pods:                pods,
-		},
-	}, nil
+// better cost and availability decisions. Pods provided are all schedulable in
+// the same zone as tightly as possible. It follows the First Fit Decreasing bin
+// packing technique, reference-
+// https://en.wikipedia.org/wiki/Bin_packing_problem#First_Fit_Decreasing_(FFD)
+func (p *podPacker) Pack(ctx context.Context, pods []*v1.Pod) ([]*Packings, error) {
+	// 1. Arrange pods in decreasing order by the amount of CPU requested, if
+	// CPU requested is equal compare memory requested.
+	sort.Sort(sort.Reverse(byResourceRequested{pods}))
+	return p.packSortedPods(pods)
+}
+
+// takes a list of pods sorted based on their resource requirements compared by CPU and memory.
+func (p *podPacker) packSortedPods(pods []*v1.Pod) ([]*Packings, error) {
+	// start with the smallest instance type and the biggest pod check how many
+	// pods can we fit, go to the next bigger type and check if we can fit more
+	// pods. Compare pods packed on all types and select the instance type with
+	// highest pod count.
+	estimator := &packingEstimator{isPodPacked: map[*v1.Pod]bool{}}
```

Comment:
I'm finding this a bit hard to read, maybe because it doesn't seem like a common pattern in Go? It seems like `packingEstimator` is being used both as a raw map that can be manipulated directly, but also as a "black box" object (like the method `calculatePackingForAllTypes`). I'd kind of expect more one or the other, either have a entirely black box object that I only use methods on, or else just use a raw map and have functions that operate on it. Happy to discuss off-line if that would be easier?